URL: Fix getQueryString incorrect handling of hash fragment

aduth opened the pull request:
Previously: #11286
Extracted from: #20693

This pull request seeks to fix an incorrect behavior of `@wordpress/url` `getQueryString`, where a result may be returned when a `?` is encountered within the hash fragment of a URL:

```js
getQueryString( 'https://example.com/#?foo' );
// Before: "foo"
// After:  undefined
```

A hash fragment is always the last segment of a URL, and the querystring must always precede it ([see reference specification](https://url.spec.whatwg.org/#absolute-url-with-fragment-string)).

**Testing Instructions:**

Ensure unit tests pass:

```
npm run test-unit packages/url
```

aduth committed:
URL: Fix getQueryString incorrect handling of hash fragment

aduth committed:
URL: Update getQueryString test cases to proper URLs

aduth commented:
As seen with the changes in 0a6046a, the new implementation is slightly more strict in regards to assuming the input argument to be a proper URL, not merely any string which happens to contain `?`. While this could limit flexibility, I believe it's an appropriate change given:

- `getQueryString` is [specifically documented to receive "the full URL"](https://github.com/WordPress/gutenberg/blob/master/packages/url/README.md#getQueryString)
- Consistency with other `@wordpress/url` functions which already use and rely on the `URL` constructor to operate against an input (e.g. #19871)

gziolo reviewed (approved):
It looks correct. I hope it doesn’t break anything in projects that depend on this bug 😅

aduth committed:
URL: Add test case for empty querystring as undefined

Previously _sort of_ existed, but as an invalid URL which would fail to construct as `new URL` in the current implementation. Now split to two assertions verifying "invalid URL" behavior and "empty querystring" behavior.

aduth commented:
> As seen with the changes in [0a6046a](https://github.com/WordPress/gutenberg/commit/0a6046a9013c5bef0456dd15cf0f0f693f747f5e), the new implementation is slightly more strict in regards to assuming the input argument to be a proper URL, not merely any string which happens to contain `?`. While this could limit flexibility, I believe it's an appropriate change given:

Note that I had to scale this back a bit in #20693. Specifically, there's at least a few occurrences where we rely on an implementation like this:

```js
addQueryArgs( '/wp-json/wp/v2/posts`, { type: 'post' } );
```

And if `@wordpress/url` only supports full URLs, it loses a lot of its usefulness.
